pkg/cache/v3/delta.go

```diff
@@ -31,10 +31,13 @@ func respondDelta(request *DeltaRequest, value chan DeltaResponse, state stream.
 	}
 
 	// Only send a response if there were changes
-	if len(resp.Resources) > 0 || len(resp.RemovedResources) > 0 {
+	if len(resp.Resources) > 0 || len(resp.RemovedResources) > 0 ||
+		// we want to respond immediatly for the first wildcard request in a stream, even if the response is empty
+		// otherwise, envoy won't complete initialization
+		(state.IsWildcard() && state.IsFresh()) {
```

Comment:
This is the only meaningful change in this PR. The rest is just refactoring: I added wrapper functions, such as `Get/SetResourceVersions` and `IsWildcard` to track whether the stream state "is fresh", which means that the state corresponds to the first request in a stream.

Reply:
Just a slight nit pick for this, can we move the `&&` block to the same line as the `if`? It's a bit hard to read like that. I'd just mention the comment is targeted at that section

Reply:
So @snowp and I originally had a bool on that Watches object that was `IsFirst`, do we really need that? Can't we just determine if there are no open watches and there's no stream state we can just assume this is the first request? That's what we currently do

Reply:
Are you referring to [this](https://github.com/envoyproxy/go-control-plane/blob/main/pkg/server/delta/v3/server.go#L164) code? We can easily infer whether this is the first request in a stream from the server code, but we can't do the same in the cache code because we don't pass watches objects to the cache. Alternatively, we can pass  isFirst flag all the way from the server, starting with CreateDeltaWatch function. But I think that the stream state should aggregate all information about the stream state, so it probably makes sense to store this information there.

Reply:
That's fair i'm in agreement with that. I'd like the stream state to contain most of the knowledge about delta streams.

I had previously tried passing down an `IsFirst()` flag down into the cache but it was awkward. Can we change the name `IsFresh()`? That seems to read a bit awkward

Reply:
Yeah, I like `IsFirst` better, renaming.

Reply:
done